### ISSUE TITLE
Change Request: Add --no-fixable-warnings option to exit with code 1

### ISSUE THREAD

**Issue by glinf** [eslint/eslint#20072]
### ESLint version

v9.34.0

### What problem do you want to solve?

ESLint currently does not provide a way to exit with code 1 if they are fixable warnings.

 I propose `--no-fixable-warnings` option in the CLI to exit with code 1

This addresses a common CI/CD inconsistency issue that causes developer frustration:

  The Problem Scenario

  Dev 1's Workflow:
  1. Writes code with some style violations (e.g., missing semicolons, wrong quotes)
  2. Runs ESLint locally → gets warnings but exit code 0 ✅
  3. Commits and pushes → CI passes ✅
  4. Thinks everything is fine

  Dev 2's Workflow:
  1. Pulls Dev 1's code
  2. Has their editor configured to auto-fix on save, OR runs eslint --fix as part of their workflow
  3. Now has "changes" in their working directory that they didn't make
  4. Has to decide: commit the fixes (polluting git history) or stash them (losing the improvements)

PR: https://github.com/eslint/eslint/pull/20071

  The Inconsistency

  - Without --no-fixable-warnings: ESLint treats fixable warnings as "acceptable" (exit code 0)
  - In practice: Many teams want fixable issues resolved immediately since they're zero-effort to fix
  - Result: Inconsistent code state across developers and CI environments

  How --no-fixable-warnings Solves This

  With this flag enabled in CI:

  Dev 1's Updated Workflow:
  1. Writes code with style violations
  2. Runs ESLint locally → gets warnings and exit code 1 ❌
  3. Cannot commit/push until running eslint --fix or manually fixing issues
  4. Commits clean, consistent code

  Dev 2's Workflow:
  1. Pulls Dev 1's code
  2. Code is already properly formatted
  3. No unexpected changes in working directory
  4. Can focus on actual development instead of formatting cleanup

### What do you think is the correct solution?

https://github.com/eslint/eslint/pull/20071

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

Naming CLI argument is hard, not attached to anything in particular but I think this is a legit use case so whatever the consensus is on the naming is fine.

**Comment by mdjermanovic** [comment]
Thanks for the detailed explanation of the use case for this option.

I think in most cases this can be solved by setting the rule severity to the `error` level, especially for mentioned formatting rules, which probably always produce fixable lint problems. However, a rule being generally fixable doesn't guarantee that all lint problems reported by it will be fixable, so this option makes sense to me 👍 

@eslint/eslint-team thoughts?

Note: I think this change would need to go through the [RFC process](https://github.com/eslint/rfcs).

**Comment by glinf** [comment]
Thanks for looking into it!

**Comment by nzakas** [comment]
I'm having a hard time understanding this use case. If you definitely want something fixed, then there are already two options:

1. Ensure the rule is set to `"error"` (ensure CI breaks)
2. Always run ESLint with `--fix` (either manually or on precommit)

I'm also not sure about using the heuristic of "fixable warnings" -- as @mdjermanovic noted, not everything that's fixable will actually be fixed. We can't guarantee that what was fixable when the code was initially committed will remain fixable once other changes are merged in.

So to me, the reasoning behind this new flag is kind of muddy. I'm not sure it's actually solving the cases listed in the OP in the way it's intended. 

That said, I don't feel strongly either way. If other team members feel this should move on to the RFC process, I'm fine with that.

**Comment by glinf** [comment]
Thanks for the feedback! Let me share my thoughts:

**Re: "just set rules to error"** - That's exactly what we *don't* want. The key distinction here is that we want these to remain warnings (not blocking development) but we want them to be fixed by the same person who commits the code, not passed along as cleanup work for the next developer.

**Re: "run eslint --fix manually/precommit"** - You've hit on exactly why this flag is needed:

1. **Manually** - This relies on developer discipline and good intentions. That's precisely how I end up picking up formatting changes in files completely unrelated to my actual work, polluting the git diff with noise.

2. **Precommit** - Sure, that's a workaround, but it's solving the problem at the wrong layer. We shouldn't need to configure additional tooling when ESLint itself can provide this functionality directly.

**Re: "not everything fixable will be fixed"** - That's a fair point about the heuristic, but in practice, the vast majority of fixable warnings do get fixed (especially the common formatting ones that cause this workflow friction). For edge cases where something is marked fixable but doesn't get fixed, developers would handle it the same way they do today. I'm flexible on the flag name, could be `--fail-on-fixable-warnings`, `--strict-fixable`, whatever the team prefers.

The core issue is that there's currently no built-in way to ensure the developer who introduces fixable issues is responsible for fixing them, rather than that work falling to whoever happens to touch the code next.

Happy to take this through the RFC process! 🙂

**Comment by fasttime** [comment]
> **Re: "just set rules to error"** - That's exactly what we _don't_ want. The key distinction here is that we want these to remain warnings (not blocking development) but we want them to be fixed by the same person who commits the code, not passed along as cleanup work for the next developer.

Could you elaborate on how ESLint errors are blocking your local development?

**Comment by glinf** [comment]
Hey @fasttime ! Sure!

Say you work on a monorepo, with a bunch of collaborators. The workflow is as follow:

```
Dev 1's Workflow:

    Writes code with some style violations (e.g., missing semicolons, wrong quotes)
    Runs ESLint locally → gets warnings but exit code 0 ✅
    Commits and pushes → CI passes ✅
    Thinks everything is fine

Dev 2's Workflow:

    Pulls Dev 1's code
    Has their editor configured to auto-fix on save, OR runs eslint --fix as part of their workflow
    Now has "changes" in their working directory that they didn't make
    Has to decide: commit the fixes (polluting git history) or stash them (losing the improvements)

```

It is not as per se "blocking" but more "interrupting", it forces the innocent Dev 2 (me) to take an annoying decision: (1) either commit the warning fix even tho it has nothing to do with the Dev 2 changes / PR (which is bad practice) or (2) ignore it and let another Dev deal with the decision (say Dev 3, poor guy will also be interrupted).

Again, I can go talk to Dev 1 and tell that Dev to run the fixes before commiting or add it to it's precommit hooks but that's relying on good intentions and we know how this goes.

**Comment by fasttime** [comment]
> Say you work on a monorepo, with a bunch of collaborators. The workflow is as follow:
> 
> ```
> Dev 1's Workflow:
> 
>     Writes code with some style violations (e.g., missing semicolons, wrong quotes)
>     Runs ESLint locally → gets warnings but exit code 0 ✅
>     Commits and pushes → CI passes ✅
>     Thinks everything is fine
> 
> Dev 2's Workflow:
> 
>     Pulls Dev 1's code
>     Has their editor configured to auto-fix on save, OR runs eslint --fix as part of their workflow
>     Now has "changes" in their working directory that they didn't make
>     Has to decide: commit the fixes (polluting git history) or stash them (losing the improvements)
> ```

Thanks! Let me rephrase the question: is there a specific reason you're using warnings instead of errors in this case? If you switched all warnings to errors, you'd likely get the desired outcome: when "Dev 1" runs ESLint locally, it would exit with code 1, causing the pre-commit hooks to fail and preventing them from committing or pushing the changes.

**Comment by glinf** [comment]
Why not just use error?

Because severity and policy aren’t the same thing. We want these rules to remain warnings (non-blocking during local iteration), while still ensuring the person who introduces fixable issues is the one who fixes them before merge. Flipping to error changes the developer experience and fails CI on all violations from that rule, including the non auto fixable ones, which is stricter than the team intends.

**Comment by DMartens** [comment]
You could also use an environment variable (e.g. `env.CI`) to set the severity of the rule, for example in `eslint.config.js`:
```javascript
const shouldErrorInCI = env.CI ? "error" : "warn";

export default {
	rules: {
		rule: shouldErrorInCI
	}
}
```

**Comment by glinf** [comment]
I know ^^ but I only want auto fixable (since it's zero overhead) to be treated as "error", not all of them.

Not every occasion for a rule means it's fixable. 

Think of the rule being: if it's zero overhead to fix, it must be fixed

**Comment by nzakas** [comment]
There's not enough interest on the team to implement this change, so closing.
